### ISSUE TITLE
Reload/navigation confirmation on PDFs with forms 

### ISSUE THREAD

**Issue by rms2219** [stephanrauh/ngx-extended-pdf-viewer#2691]
When there is a form on a PDF, and the user interacts with the PDF, they will _always_ be prompted if they haven't downloaded the PDF prior to reloading or navigating away regardless of any form values changing.

- Does the bug also happen on https://mozilla.github.io/pdf.js/web/viewer.html? (If so, please open the issue at Mozilla's bugtracker).

Maybe? There is no form on this sample PDF so I'm not sure if it's an issue w/ the library or not.

I opened the PDF in Firefox, which uses the Mozilla PDF.js, and the issue does not appear there.

- Does the bug also happen on https://pdfviewer.net/simple?

Yes

**Describe the bug**
1. Load any PDF that has a form on it (use default options)
2. Click on a form field, but make no changes/edits
4. Reload or navigate away from page
5. Observe confirmation notifying that changes have not been saved.

No changes have actually been made, so there should be no prompt. Additional unrelated changes (e.g. change the zoom of the PDF) will also trigger this same condition.

Ideally, there should be an option that disables any such prompt, without explicitly disabling the form. For our use case, we have no need to ensure the user wants to save any changes to the PDF. Our users will never be editing forms anyway, but we don't want to explicitly disable the form in the off-chance that they would like to edit and download a local copy. It should be noted that even using `[disableForms]` option still presents this confirmation.

**Version info**
- 22.0.1

Example PDF: https://www.signify.com/api/assets/v1/file/Signify/content/a654757c91404dbf8360ac8b0104206e/Swan-LED-Pendant--CXF4----Spec-Sheet--G3-.pdf

**Comment by rms2219** [comment]
To add to this, I found that if I handle the `onPageRendered` event, I can do this:

`pageView.annotationLayer.annotationStorage.resetModified();`

This allows me to proceed without the prompt to save changes. If I do, in fact, set a form field and try to reload, I'll still be prompted.

**Comment by stephanrauh** [comment]
The key point is that the - usually desired - confirmation dialog appears when you click a field without changing anything.

At the moment, I can't say much about it. As far as I can see, that doesn't happen on https://pdfviewer.net/extended-pdf-viewer/forms. Can you confirm that? If so, the problem might be your PDF file. It's possible that the file contains some magic that makes the PDF viewer (and the browser) believe you've edited the form.

Note that this is merely a first assessment. I haven't examined your PDF file yet (it's late in the evening), but maybe me thoughts already help you.

**Comment by rms2219** [comment]
> Can you confirm that?

I cannot confirm that. In your forms sample, simply loading the page, scrolling the PDF and attempting to reload will cause this confirmation dialog. I did not interact with a form element at all. This was done in Firefox.

I've attached a screen recording for you to see my actions.

https://github.com/user-attachments/assets/62860b55-7af5-457a-a751-f766718f22bf

As an aside, I'm no longer to load that example page in Brave, yet I could yesterday:

<img width="667" alt="image" src="https://github.com/user-attachments/assets/b4c06144-8cca-483c-8612-7430dcd61350" />

**Comment by rms2219** [comment]
@stephanrauh, any updates on this?

**Comment by stephanrauh** [comment]
Thanks for reminding me! I think I've finally understood and reproduced the issue.

**Comment by stephanrauh** [comment]
I've got an idea how to solve this issue - but it's a bit convoluted. Calling `resetModified()` is a clever solution, but I'm not convinced we should call it when a page has been rendered. Large PDF files are rendered lazily, so you might loose real changes, and the pages are also rendered if the user zooms.

**Comment by rms2219** [comment]
> Calling `resetModified()` is a clever solution, but I'm not convinced we should call it when a page has been rendered.

Yeah, I wouldn't expect that to be the solution, though I wanted to point out that it "worked" for me when reproducing the bug if for no other reason than to possibly direct you service/object responsible for the bug.

**Comment by stephanrauh** [comment]
Yes, and I appreciate your did - you're pointing into the right direction!

Thing is, I believe this is a peculiarity of the base library. I didn't manage to reproduce it on https://mozilla.github.io/pdf.js/web/viewer.html, but still, I suspect it's one of the "deep" problems. I don't mind modifying pdf.js if there's a good reason and if I can be sure there's little-to-none long-term risk, but I prefer to be careful. Every modification I do makes the next merge request more challenging.

So I'm looking for a minimal-invasive solution. That might involve storing the state of every field immediately after loading and resetting the state on every event like ´(input)`, `(blur)` if there's no real change. That feels a bit messy. On the other hand, it might even be beneficial: if the user modifies a field and reverts it to the original state, the new approach allows the library to detect nothing has changed.

**Comment by stephanrauh** [comment]
BTW, I expect to be short of time during the next days or weeks - but you've successfully piqued my curiosity!

**Comment by rms2219** [comment]
> Thing is, I believe this is a peculiarity of the base library.

Perhaps through your exploration, you'll prove that and we can get it fixed in the right spot without having to modify your fork.

> I didn't manage to reproduce it on https://mozilla.github.io/pdf.js/web/viewer.html

Is this even a PDF that would show the error? In my experience, it's only on PDFs with form elements, and at first glance, I didn't see any on that PDF. That said, opening my sample PDF in Firefox _also_ did not reproduce the issue.

**Comment by stephanrauh** [comment]
Well, the pdf.js showcase - https://mozilla.github.io/pdf.js/web/viewer.html - does have an "open file" menu. It's just hidden in the secondary menu on the right-hand side. In hindsight, I suspect I didn't get you right when I tried to reproduce the error. It should show on the pdf.js showcase, too.

**Comment by rms2219** [comment]
> Well, the pdf.js showcase - https://mozilla.github.io/pdf.js/web/viewer.html - does have an "open file" menu. 

Who knew?!   :) 

> In hindsight, I suspect I didn't get you right when I tried to reproduce the error. It should show on the pdf.js showcase, too.

It only happens on the showcase when I make the a form field "dirty". If I simply load the PDF or give focus to a field (but don't set a value), and navigate away, I'm not prompted. That's not the case when using your library.

**Comment by stephanrauh** [comment]
My idea is to allow the user to change a field, revert the change, and then the library shouldn't annoy them with the confirmation dialog. Probably it'll take a while to find the correct implementation because there's also `[formData]`, and I assume that the initial value of `[formData]` should be the baseline.

At the moment, I'm short of time, but don't worry: I've started to work on your issue. It made it to the top of the pile on my desktop!

**Comment by stephanrauh** [comment]
I've solved your issue, it just needs some additional testing. At the moment I'm reluctant to publish a new version. Maybe you've heard about the "Shai-Hulud" attack on npm. I'm 99% sure my computer is clean - but I'd prefer to wait a while until the dust has settled before publishing a new version. At the moment, CISA warns developers to install packages that have been published after September, 16th. If I'm publishing a new version now, that'd fall under that category, making developers ignore the version.

**Comment by stephanrauh** [comment]
The bug fix has landed with version 26.0.0-alpha.25.
